jupyter_publishing_service/storage/base.py

```diff
@@ -0,0 +1,166 @@
+        self.user_store = self.user_store_class(parent=self, log=self.log)
+
+    async def start(self):
+        # Don't do anything here. Subclasses can use this to initialize a e.g. database.
+        ...
+
+    async def authorize(self, user: Collaborator, file_id: str) -> bool:
+        return await self.authorization_store.authorize(user, file_id)
+
+    async def get(
+        self, file_id: str, collaborators: bool = False, contents: bool = False
+    ) -> SharedFileResponseModel:
+        metadata: SharedFileMetadata = await self.metadata_store.get(file_id)
+        collaborator_roles = None
+        if collaborators:
+            collaborator_roles: List[CollaboratorRole] = await self.collaborator_store.get(file_id)
+        file = None
+        if contents:
+            file: JupyterContentsModel = await self.file_store.get(file_id=file_id)
+        return SharedFileResponseModel(
+            metadata=metadata, collaborator_roles=collaborator_roles, contents=file
+        )
+
+    async def add(self, request_model: SharedFileRequestModel) -> SharedFileResponseModel:
+        """
+        Store a new shared file.
+
+        Returns a SharedFileResponse without contents and collaborators.
+        """
+        metadata = await self.metadata_store.add(request_model.metadata)
+        if request_model.collaborators:
+            for collaborator in request_model.collaborators:
+                # The author should have writer permissions.
+                if collaborator.email == metadata.author:
+                    await self.collaborator_store.add(
+                        request_model.metadata.id, collaborator, [Role(name="WRITER")]
+                    )
+                    continue
+                await self.collaborator_store.add(
+                    request_model.metadata.id, collaborator, request_model.roles
+                )
+        if request_model.contents:
+            await self.file_store.add(metadata.id, request_model.contents)
+        return SharedFileResponseModel(metadata=metadata)
+
+    async def delete(self, file_id: str):
+        # Need to remove collaborators for this file.
+        collaborator_roles = await self.collaborator_store.get(file_id=file_id)
+        # NOTE: we should refactor this to delete as a batch, not one-by-one.
+        for cr in collaborator_roles:
+            await self.collaborator_store.delete(file_id, Collaborator(email=cr.email))
+        # Delete file and metadata
+        await self.file_store.delete(file_id)
+        await self.metadata_store.delete(file_id)
+
+    async def update(
+        self, file_id: str, request_model: SharedFileRequestModel
+    ) -> SharedFileResponseModel:
+        metadata = await self.metadata_store.update(request_model.metadata)
+        if request_model.collaborators:
+            await self.collaborator_store.update(
+                file_id, request_model.collaborators, request_model.roles
```

Comment:
The update interface takes only a single collaborator and multiple roles. Did you mean for it to accept multiple collaborators as is the type inferred from request_model.collaborators.

Reply:
I wrote this to make the REST API and storage manager interface as general as possible. Why couldn't we have multiple collaborators and multiple roles updated in a single request?

Reply:
Let me test this out in parallel with the frontend extension and return to this later?